src/yachtzee/diet.ts

```diff
@@ -0,0 +1,692 @@
+): Array<YachtzeeDietEntry<void>> {
+  const dietSchedule = new Array<YachtzeeDietEntry<void>>();
+  const remainingMenu = new Array<YachtzeeDietEntry<void>>();
+  const jellies = new Array<YachtzeeDietEntry<void>>();
+  const haveDistentionPill = !get("_distentionPillUsed") && have($item`distention pill`);
+  const toasts = new Array<YachtzeeDietEntry<void>>();
+  const freeNCs = new Array<YachtzeeDietEntry<void>>();
+
+  // We assume the menu was constructed such that we will not overshoot our fullness and inebriety limits
+  // Assume all fullness/drunkenness > 0 non-spleen cleansers are inserted for buffs
+  // This makes it trivial to plan the diet
+  // First, lay out all the spleen cleansers (and the buff consumables at the front)
+  for (const entry of menu) {
+    if (entry.spleen < 0) {
+      for (const splitEntry of splitDietEntry(entry)) dietSchedule.push(splitEntry);
+    } else if (entry.name === "stench jelly") {
+      for (const splitEntry of splitDietEntry(entry)) jellies.push(splitEntry);
+    } else if (entry.name === "toast with stench jelly") {
+      for (const splitEntry of splitDietEntry(entry)) toasts.push(splitEntry);
+    } else if (["clara's bell", "jurassic parka"].includes(entry.name)) {
+      for (const splitEntry of splitDietEntry(entry)) freeNCs.push(splitEntry);
+    } else if (entry.fullness > 0 || entry.drunkenness > 0) {
+      for (const splitEntry of splitDietEntry(entry)) dietSchedule.splice(0, 0, splitEntry);
+    } else {
+      for (const splitEntry of splitDietEntry(entry)) remainingMenu.push(splitEntry);
+    }
+  }
+
+  // Place toasts at the back of our schedule
+  for (const toast of toasts) {
+    dietSchedule.push(toast);
+  }
+
+  // Then, greedily inject spleen items into the schedule with the ordering:
+  // 1) Front to back of the schedule
+  // 2) Large spleen damagers to small spleen damagers
+  // This works because stench jellies are of size 1, so we can always pack efficiently using the greedy approach
+  remainingMenu.sort((left, right) => {
+    return right.spleen - left.spleen;
+  });
+
+  // Schedule jellies last so we definitely get spleen buffs first (e.g. synth and cologne)
+  for (const spleeners of [remainingMenu, jellies]) {
+    for (const entry of spleeners) {
+      let idx = 0;
+      let spleenUse = mySpleenUse();
+      while (
+        idx < dietSchedule.length &&
+        (dietSchedule[idx].spleen >= 0 || // We only insert if there's a cleanser immediately after where we want to insert
+          spleenUse + entry.spleen > spleenLimit() || // But don't insert if we will overshoot our spleen limit
+          (idx > 0 &&
+            dietSchedule[idx - 1].spleen < 0 &&
+            spleenUse + dietSchedule[idx].spleen >= 0)) // And cluster spleen cleansers (continue if the next cleanser can still clean our spleen)
+      ) {
+        spleenUse += dietSchedule[idx++].spleen ?? 0;
+      }
+      dietSchedule.splice(idx, 0, entry);
+    }
+  }
+
```

Comment:
To support `sweat out some booze`, we can use the same loop as the overconsumption check below. The moment our inebriety is larger than the number of sweats `(x)` we get to use, insert `x` sweat outs

Reply:
Out of scope of this PR